### PR TITLE
while_immutable_condition: fix handling of self

### LINKED ISSUE #2540: false positive on while_immutable_condition
Run clippy on [dimbleby/mio-extras](https://github.com/dimbleby/mio-extras), get this warning:

```
$ cargo +nightly clippy
lib: mio-extras
   Compiling cfg-if v0.1.2
   Compiling libc v0.2.39
   Compiling lazycell v0.6.0
   Compiling slab v0.4.0
   Compiling log v0.4.1
   Compiling net2 v0.2.32
   Compiling iovec v0.1.2
   Compiling mio v0.6.14
   Compiling mio-extras v2.0.3 (file:///data/dch/mio-extras)
warning: Variable in the condition are not mutated in the loop body. This either leads to an infinite or to a never running loop.
   --> src/timer.rs:266:9
    |
266 | /         while self.tick <= target_tick {
267 | |             let curr = self.next;
268 | |
269 | |             trace!("ticking; curr={:?}", curr);
...   |
307 | |             }
308 | |         }
    | |_________^
    |
    = note: #[warn(while_immutable_condition)] on by default
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.188/index.html#while_immutable_condition

    Finished dev [unoptimized + debuginfo] target(s) in 4.28 secs
```

This is incorrect - the loop progresses [here](https://github.com/dimbleby/mio-extras/blob/3341fa96741c214e8b79500e96606a6b83192180/src/timer.rs#L272).

```
$ cargo +nightly clippy --version
0.0.188
```

### PR DESCRIPTION
**Pull request by kimsnj**
Fixes #2540 

`self` mutation was not properly handled for two reasons:
-  When collection variable used in the while condition and [checking for mutability](https://github.com/rust-lang-nursery/rust-clippy/blob/eafd09010815da43302ac947afee45b0f5219e6b/clippy_lints/src/loops.rs#L1482), the annotation of the `Def` was set to No for `self` even when mutable.
- The `ExprUseVisitor` calls the [`Delegate` callback](https://github.com/rust-lang-nursery/rust-clippy/blob/eafd09010815da43302ac947afee45b0f5219e6b/clippy_lints/src/loops.rs#L2216) with categorization `Deref` and not directly `Local`.

With this fix, variables of the condition are checked for mutation by their usage and not by their definition. We thus cannot have the message *"all variables in condition are immutable"* anymore.

### CODE REVIEW

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -2208,15 +2210,18 @@ impl<'a, 'tcx> Visitor<'tcx> for MutableVarsVisitor<'a, 'tcx> {
...
 impl<'tcx> MutVarsDelegate {
     fn update(&mut self, cat: &'tcx Categorization, sp: Span) {
-        if let Categorization::Local(id) = *cat {
-            if let Some(span) = self.mut_spans.get_mut(&id) {    
-                *span = Some(sp)
-            }
+        match *cat {
+            Categorization::Local(id) =>
+                if let Some(span) = self.mut_spans.get_mut(&id) {
+                    *span = true;
```
Comment (oli-obk) [inline clippy_lints/src/loops.rs]: Why can't we keep spans here? I think it would be great if we pointed to the variable instead of the entire loop

Comment (kimsnj) [reply clippy_lints/src/loops.rs]: Here the spans are for the variables that actually get mutated. So as soon as one is set, we do not lint. I figured a simple bool would better fit the usage .
(but I should indeed rename the field if we stick to a map from id to bool…)

Comment (kimsnj) [reply clippy_lints/src/loops.rs]: Sure! Updated :)

Comment (oli-obk) [reply clippy_lints/src/loops.rs]: Thanks!

**Pull request merged**